src/csharp/Grpc.Core/GrpcEnvironment.cs

```diff
@@ -360,12 +361,25 @@ internal static string GetCoreVersionString()
 
         internal static void GrpcNativeInit()
         {
+            if (!IsNativeShutdownAllowed && nativeInitCounter.Count > 0)
+            {
+                // Normally grpc_init and grpc_shutdown calls should come in pairs (C core does reference counting),
+                // but in case we avoid grpc_shutdown calls altogether, calling grpc_init has no effect
+                // besides incrementing an internal C core counter that could theoretically overflow.
+                // To avoid this theoretical possibility we guard repeated calls to grpc_init()
+                // with a 64-bit atomic counter (that can't realistically overflow).
+                return;
+            }
             NativeMethods.Get().grpcsharp_init();
+            nativeInitCounter.Increment();
```

Comment:
@apolcyn  because the GrpcNativeInit() is invoked for every channel creation / shutdown and server creation / shutdown, I wanted to avoid a global lock, so I used an AtomicCounter, to determine if grpcsharp_init needs to be invoked.
PTAL.